### ISSUE TITLE
"No scope" warning for pull request closed events

### ISSUE THREAD

**Issue by SatheeshJM** [actions/cache#129]
I get the warning `No scopes with read permission were found on the request.` if I setup my workflow for triggering on pull request: close.  
If I change the trigger to push, it works. 

Is the problem that the head_branch of the Pull Request can't read/write to the master's cache? 

An example config

```
on:
  pull_request:
    types: 
      - closed
    branches: 
      - master

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
      with:
        ref: master
        fetch-depth: 1
    - name: Cache node modules
      id: cache_node_modules
      uses: actions/cache@v1
      with: 
        path: node_modules
        key: ${{ runner.os }}-node-${{ hashFiles('package.json') }}
```
for the above config the cache is never used.. I get the below warning
![image](https://user-images.githubusercontent.com/1222562/70833067-d3d39d80-1e1c-11ea-80fa-5eb3f56fce3e.png)

If I tweak the workflow to trigger on branch push, all is well
```
on:
  push:
    branches: 
      - master

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/checkout@v2
      with:
        ref: master
        fetch-depth: 1
    - name: Cache node modules
      id: cache_node_modules
      uses: actions/cache@v1
      with: 
        path: node_modules
        key: ${{ runner.os }}-node-${{ hashFiles('package.json') }}
```
![image](https://user-images.githubusercontent.com/1222562/70833159-0f6e6780-1e1d-11ea-8def-80c5d6712054.png)

**Comment by joshmgross** [comment]
>Is the problem that the head_branch of the Pull Request can't read/write to the master's cache?

Yep! You'll need to trigger off of a `push` event to write to the default branch's cache.

**Comment by SatheeshJM** [comment]
I need my action to trigger on a `pull_request` event.

Restricting write makes sense.  But reads too? 
From the docs
```
Example of search priority
key:
  npm-feature-d5ea0750
restore-keys: |
  npm-feature-
  npm-
For example, if a pull request contains a feature branch (the current scope) and targets the default branch (master), the action searches for key and restore-keys in the following order:

Key npm-feature-d5ea0750 in the feature branch scope
Key npm-feature- in the feature branch scope
Key npm- in the feature branch scope
Key npm-feature-d5ea0750 in the master branch scope
Key npm-d5ea0750 in the master branch scope
Key npm in the master branch scope
```
Doesn't ☝️ imply that a feature branch would be able to read cache which is in the scope of the master branch?

**Comment by SatheeshJM** [comment]
Even if I save master cache during a push event (like mentioned here https://github.com/actions/cache/issues/79#issuecomment-552246833), my pull event is not able to read the master cache.

**Comment by chemitaxis** [comment]
Can you add @joshmgross to the documentation what events have scope? On "deployment" event does not work either...

**Comment by SatheeshJM** [comment]
@joshmgross any thoughts? 
Essentially my branch is not able to access the default branch scoped cache

**Comment by joshmgross** [comment]
@SatheeshJM the issue is a mix of scopes and how we determine which events to give write and read access.
For PRs, we validate that the ref is in the form that we expect `refs/pulls...` (i.e. it's a pull request branch). Since this is a PR Close event and you're not seeing cache permissions, I'd expect this ref to be in some other form.

>Can you add @joshmgross to the documentation what events have scope? On "deployment" event does not work either...

Documentation is here: https://help.github.com/en/actions/automating-your-workflow-with-github-actions/caching-dependencies-to-speed-up-workflows#restrictions-for-accessing-a-cache
>You can only access the cache in workflows triggered by push and pull_request events.

We may have to modify to specify which types of `pull_request` events are supported.

>On "deployment" event does not work either...
Tracking that with https://github.com/actions/cache/issues/63

**Comment by SatheeshJM** [comment]
So are you saying the default branch's cache will be accessible from other PR events, but PR-close event won't have access, because the ref will be different? 
@joshmgross

**Comment by joshmgross** [comment]
@SatheeshJM that's the current behavior yes.

**Comment by Duckle29** [comment]
An update of the docs for that would be nice, as this is currently the only place this behavior is documented :)

**Comment by klapec** [comment]
I'm getting the same error when doing a re-run by clicking the "re-run all jobs" button after a failed run. Is there anything I can do here?

**Comment by joshmgross** [comment]
> An update of the docs for that would be nice, as this is currently the only place this behavior is documented :)

Yep we can get this documented.


>I'm getting the same error when doing a re-run by clicking the "re-run checks" button after a failed run. Is there anything I can do here?

What event was your original run? Did it work with caching?
If you have a link to the run that would be helpful

**Comment by klapec** [comment]
I can't share the link directly as it is a private repo, but here are the logs for the job that fetches the NPM deps and caches them: https://gist.github.com/klapec/3d62182192afdb29d72142a008b95a02
and here are the logs for the job that's supposed to build the application (and it obviously fails because the dependencies were not loaded from the cache):
https://gist.github.com/klapec/41a988381d1f58e2baa727eca51e2b6a

**Comment by daniel-white** [comment]
@joshmgross why is a read forbidden? its not mutating things. i have an upstream provider that has been really flaky as of late and the cache saves me lots of breaks.

**Comment by jcornaz** [comment]
So if I understand well we cannot use `actions/cache` for pull requests?

**Comment by joshmgross** [comment]
>So if I understand well we cannot use actions/cache for pull requests?

No that is not correct. All `pull_request` events work except for the `closed` type. By default, this event will not trigger workflows. See https://help.github.com/en/actions/reference/events-that-trigger-workflows#pull-request-event-pull_request

**Comment by jcornaz** [comment]
Ok. Thanks @joshmgross.

So if all `pull_request` events are supposed to work, there is a problem. I opened an issue for it: #208.

**Comment by joshmgross** [comment]
The issue of pull request re-runs has been fixed, see #208 for details.

With V2 of this action, all events with a GitHub ref are supported, so this issue should now be fixed. See https://github.com/actions/cache#whats-new for more info.
